libpod/container_api.go

```diff
@@ -125,84 +166,80 @@ func (c *Container) Start(ctx context.Context) (err error) {
 			return err
 		}
 	}
-
-	// Start the container
-	return c.start()
+	return nil
 }
 
-// StartAndAttach starts a container and attaches to it
-// StartAndAttach can start configured, created or stopped containers
-// For configured containers, the container will be initialized first, then
-// started
-// Stopped containers will be deleted and re-created in runc, undergoing a fresh
-// Init()
-// If successful, an error channel will be returned containing the result of the
-// attach call.
-// The channel will be closed automatically after the result of attach has been
-// sent
-func (c *Container) StartAndAttach(ctx context.Context, streams *AttachStreams, keys string, resize <-chan remotecommand.TerminalSize) (attachResChan <-chan error, err error) {
+// RestartWithTimeout restarts a running container and takes a given timeout in uint
+func (c *Container) RestartWithTimeout(ctx context.Context, timeout uint) (err error) {
 	if !c.batched {
 		c.lock.Lock()
 		defer c.lock.Unlock()
 
 		if err := c.syncContainer(); err != nil {
-			return nil, err
+			return err
 		}
 	}
 
-	// Container must be created or stopped to be started
-	if !(c.state.State == ContainerStateConfigured ||
-		c.state.State == ContainerStateCreated ||
-		c.state.State == ContainerStateStopped ||
-		c.state.State == ContainerStateExited) {
-		return nil, errors.Wrapf(ErrCtrStateInvalid, "container %s must be in Created or Stopped state to be started", c.ID())
+	if err = c.checkDependenciesAndHandleError(ctx); err != nil {
+		return err
 	}
 
+	return c.restartWithTimeout(ctx, timeout)
+}
+
+// helper function that handles dependencies and prints a helpful message
+func (c *Container) checkDependenciesAndHandleError(ctx context.Context) error {
```

Comment:
Helper methods into container_internal.go please - we try to keep only public API stuff in container_api.go